In [1]:
# try using python-twitter: 
import twitter

In [3]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
%matplotlib notebook
plt.style.use('ggplot')

In [4]:
with open('twitterkeys.txt') as f:
    consumer_key = f.readline().strip()
    consumer_secret = f.readline().strip()
with open('twittertokens.txt') as f:
    access_token = f.readline().strip()
    access_token_secret = f.readline().strip()

In [7]:
api = twitter.Api(consumer_key=consumer_key,
                      consumer_secret=consumer_secret,
                      access_token_key=access_token,
                      access_token_secret=access_token_secret)

In [8]:
print(api.VerifyCredentials())

{"created_at": "Tue May 20 03:17:58 +0000 2008", "description": "Data scientist, vegan, feminist, lover of technology & funny/cute things. Machine learning FTW", "favourites_count": 934, "followers_count": 493, "friends_count": 1068, "geo_enabled": true, "id": 14841348, "lang": "en", "listed_count": 52, "location": "Seattle", "name": "Emily Y Spahn", "profile_background_color": "EDECE9", "profile_background_image_url": "http://abs.twimg.com/images/themes/theme3/bg.gif", "profile_image_url": "http://pbs.twimg.com/profile_images/609773934817378304/LAK6ijYl_normal.jpg", "profile_link_color": "088253", "profile_sidebar_fill_color": "E3E2DE", "profile_text_color": "634047", "screen_name": "eyspahn", "status": {"created_at": "Mon Oct 17 03:08:15 +0000 2016", "hashtags": [], "id": 787852699844751360, "id_str": "787852699844751360", "lang": "en", "media": [{"display_url": "pic.twitter.com/wEVEA6KzHP", "expanded_url": "https://twitter.com/BenedictEvans/status/787833598141313024/photo/1", "id": 

In [13]:
api.GetUserTimeline?

In [15]:
list = api.GetUserTimeline(screen_name='realDonaldTrump', since_id=None, max_id=None, count=None, 
                    include_rts=True, trim_user=False, exclude_replies=False)

In [17]:
list[0]

Status(ID=788552360326893568, ScreenName=realDonaldTrump, Created=Wed Oct 19 01:28:27 +0000 2016, Text=u'\u2018Food Groups\u2019 \u2013 Emails Show Clinton Campaign Organized Potential VPs By Race And Gender: https://t.co/Qk1fOm1t8L')

### Booo, do I have to do extra work to get the device type? -- need android for trump.
Fine, I'll go back to tweepy, I guess...

In [19]:
import tweepy

In [20]:
# Create an authentication using the given key and secret
auth = tweepy.OAuthHandler(consumer_key=consumer_key, 
    consumer_secret=consumer_secret)

# Connect to the Twitter API using the authentication
api = tweepy.API(auth)

In [24]:
firstbatch = api.user_timeline(screen_name='realDonaldTrump')

In [26]:
len(firstbatch)

20

In [28]:
# look at the various parameters from the batch of Status objects
dir(firstbatch[0])

['__class__',
 '__delattr__',
 '__dict__',
 '__doc__',
 '__eq__',
 '__format__',
 '__getattribute__',
 '__getstate__',
 '__hash__',
 '__init__',
 '__module__',
 '__ne__',
 '__new__',
 '__reduce__',
 '__reduce_ex__',
 '__repr__',
 '__setattr__',
 '__sizeof__',
 '__str__',
 '__subclasshook__',
 '__weakref__',
 '_api',
 '_json',
 'author',
 'contributors',
 'coordinates',
 'created_at',
 'destroy',
 'entities',
 'favorite',
 'favorite_count',
 'favorited',
 'geo',
 'id',
 'id_str',
 'in_reply_to_screen_name',
 'in_reply_to_status_id',
 'in_reply_to_status_id_str',
 'in_reply_to_user_id',
 'in_reply_to_user_id_str',
 'is_quote_status',
 'lang',
 'parse',
 'parse_list',
 'place',
 'possibly_sensitive',
 'retweet',
 'retweet_count',
 'retweeted',
 'retweets',
 'source',
 'source_url',
 'text',
 'truncated',
 'user']

In [29]:
firstbatch[0]

Status(contributors=None, truncated=False, text=u'\u2018Food Groups\u2019 \u2013 Emails Show Clinton Campaign Organized Potential VPs By Race And Gender: https://t.co/Qk1fOm1t8L', is_quote_status=False, in_reply_to_status_id=None, id=788552360326893568, favorite_count=9970, _api=<tweepy.api.API object at 0x114cc28d0>, author=User(follow_request_sent=None, has_extended_profile=False, profile_use_background_image=True, _json={u'follow_request_sent': None, u'has_extended_profile': False, u'profile_use_background_image': True, u'default_profile_image': False, u'id': 25073877, u'profile_background_image_url_https': u'https://pbs.twimg.com/profile_background_images/530021613/trump_scotland__43_of_70_cc.jpg', u'verified': True, u'translator_type': u'none', u'profile_text_color': u'333333', u'profile_image_url_https': u'https://pbs.twimg.com/profile_images/1980294624/DJT_Headshot_V2_normal.jpg', u'profile_sidebar_fill_color': u'C5CEC0', u'entities': {u'url': {u'urls': [{u'url': u'https://t.co/

In [36]:
firstbatch[10].source

u'Twitter for iPhone'

In [45]:
firstbatch[0].parse_list?

Object `parse_list` not found.


In [ ]:
firstbatch[0].parse_list

In [46]:
# #pagination: returns Status objects in batches of 20
# # up to 3220 total tweets at a time.
## beyond that, use the options for max_id, since_id

# for page in tweepy.Cursor(api.user_timeline, screen_name='realDonaldTrump').pages():
#     # page is a list of statuses
#     process_page(page)

tweets=[]
for page in tweepy.Cursor(api.user_timeline, screen_name='realDonaldTrump').pages():
    tweets.append(page)

In [59]:
# Need to unpack tweets from lists in lists
unpacked_tweets = [status for tweet in tweets for status in tweet]

In [62]:
len(unpacked_tweets)

3210

In [64]:
print unpacked_tweets[0].id
print unpacked_tweets[3209].id

788552360326893568
691992417336172544


In [77]:
# Next batch! But, it's not needed!
tweets = []
for page in tweepy.Cursor(api.user_timeline, screen_name='realDonaldTrump',
                          max_id=691992417336172544).pages():
    tweets.append(page)



In [81]:
tweets[0]

[Status(contributors=None, truncated=False, text=u'"@xXFlame: .@bobvanderplaats Idiot Bob Plaats on CNN is DELUDED! Cruz\'s favorability ratings are 0%, NO ONE LIKES HIM. VOTE TRUMP INSTEAD!"', is_quote_status=False, in_reply_to_status_id=None, id=691992417336172544, favorite_count=3056, _api=<tweepy.api.API object at 0x114cc28d0>, author=User(follow_request_sent=None, has_extended_profile=False, profile_use_background_image=True, _json={u'follow_request_sent': None, u'has_extended_profile': False, u'profile_use_background_image': True, u'default_profile_image': False, u'id': 25073877, u'profile_background_image_url_https': u'https://pbs.twimg.com/profile_background_images/530021613/trump_scotland__43_of_70_cc.jpg', u'verified': True, u'translator_type': u'none', u'profile_text_color': u'333333', u'profile_image_url_https': u'https://pbs.twimg.com/profile_images/1980294624/DJT_Headshot_V2_normal.jpg', u'profile_sidebar_fill_color': u'C5CEC0', u'entities': {u'url': {u'urls': [{u'url': u

In [85]:
unpacked_tweets[3209].created_at

datetime.datetime(2016, 1, 26, 14, 33, 43)

In [86]:
tw = unpacked_tweets

In [88]:
dir(tw[0])

['__class__',
 '__delattr__',
 '__dict__',
 '__doc__',
 '__eq__',
 '__format__',
 '__getattribute__',
 '__getstate__',
 '__hash__',
 '__init__',
 '__module__',
 '__ne__',
 '__new__',
 '__reduce__',
 '__reduce_ex__',
 '__repr__',
 '__setattr__',
 '__sizeof__',
 '__str__',
 '__subclasshook__',
 '__weakref__',
 '_api',
 '_json',
 'author',
 'contributors',
 'coordinates',
 'created_at',
 'destroy',
 'entities',
 'favorite',
 'favorite_count',
 'favorited',
 'geo',
 'id',
 'id_str',
 'in_reply_to_screen_name',
 'in_reply_to_status_id',
 'in_reply_to_status_id_str',
 'in_reply_to_user_id',
 'in_reply_to_user_id_str',
 'is_quote_status',
 'lang',
 'parse',
 'parse_list',
 'place',
 'possibly_sensitive',
 'retweet',
 'retweet_count',
 'retweeted',
 'retweets',
 'source',
 'source_url',
 'text',
 'truncated',
 'user']

In [127]:
def parse_tweets(tweet):
    return [tweet.created_at, tweet.id, tweet.text, tweet.source]


In [105]:
trumptweetsarr = []

for tweet in tw:
    trumptweetsarr.append(parse_trump_tweets(tweet))
    

In [106]:
trumptweetsarr

[[datetime.datetime(2016, 10, 19, 1, 28, 27),
  788552360326893568,
  u'\u2018Food Groups\u2019 \u2013 Emails Show Clinton Campaign Organized Potential VPs By Race And Gender: https://t.co/Qk1fOm1t8L',
  u'Twitter for iPhone'],
 [datetime.datetime(2016, 10, 19, 1, 0, 40),
  788545364664070144,
  u'#DrainTheSwamp \nhttps://t.co/wqqPjxfBoJ',
  u'Twitter for iPhone'],
 [datetime.datetime(2016, 10, 19, 0, 53, 43),
  788543616926912512,
  u'Time to #DrainTheSwamp in Washington, D.C. and VOTE #TrumpPence16 on 11/8/2016. Together, we will MAKE AMERICA SAFE\u2026 https://t.co/rVcjXdWxzp',
  u'Twitter for iPhone'],
 [datetime.datetime(2016, 10, 19, 0, 34, 21),
  788538743950938112,
  u'Hillary is the most corrupt person to ever run for the presidency of the United States. #DrainTheSwamp https://t.co/xA3YO8YzCq',
  u'Twitter for iPhone'],
 [datetime.datetime(2016, 10, 19, 0, 31, 25),
  788538003836678144,
  u"'Clinton Campaign Tried to Limit Damage From Classified Info on Email Server' #DrainThe

In [111]:
trumptweetsdf = pd.DataFrame(trumptweetsarr, columns = ['datetime', 'id', 'text', 'source'])

In [112]:
trumptweetsdf.head()

datetime                  id  \
0 2016-10-19 01:28:27  788552360326893568   
1 2016-10-19 01:00:40  788545364664070144   
2 2016-10-19 00:53:43  788543616926912512   
3 2016-10-19 00:34:21  788538743950938112   
4 2016-10-19 00:31:25  788538003836678144   

                                                text              source  
0  ‘Food Groups’ – Emails Show Clinton Campaign O...  Twitter for iPhone  
1           #DrainTheSwamp \nhttps://t.co/wqqPjxfBoJ  Twitter for iPhone  
2  Time to #DrainTheSwamp in Washington, D.C. and...  Twitter for iPhone  
3  Hillary is the most corrupt person to ever run...  Twitter for iPhone  
4  'Clinton Campaign Tried to Limit Damage From C...  Twitter for iPhone

In [114]:
# Use the republican convention as the start date - July 19th
trumptweetsdf.dtypes

datetime    datetime64[ns]
id                   int64
text                object
source              object
dtype: object

In [117]:
trumpdf = trumptweetsdf[trumptweetsdf.datetime>'2016-07-19 00:00:00']

In [119]:
trumpdf.source.value_counts

<bound method Series.value_counts of 0        Twitter for iPhone
1        Twitter for iPhone
2        Twitter for iPhone
3        Twitter for iPhone
4        Twitter for iPhone
5        Twitter for iPhone
6        Twitter for iPhone
7        Twitter for iPhone
8        Twitter for iPhone
9        Twitter for iPhone
10       Twitter for iPhone
11       Twitter for iPhone
12       Twitter for iPhone
13       Twitter for iPhone
14      Twitter for Android
15       Twitter for iPhone
16       Twitter for iPhone
17       Twitter for iPhone
18       Twitter for iPhone
19       Twitter for iPhone
20       Twitter for iPhone
21       Twitter for iPhone
22       Twitter for iPhone
23       Twitter for iPhone
24       Twitter Web Client
25      Twitter for Android
26       Twitter for iPhone
27       Twitter for iPhone
28       Twitter for iPhone
29      Twitter for Android
               ...         
1041       Twitter for iPad
1042    Twitter for Android
1043    Twitter for Android
1044    Twi

In [124]:
trumpdf = trumpdf[trumpdf.source=='Twitter for Android']

In [173]:
trumpdf = trumpdf.reset_index(drop=True)

In [174]:
len(trumpdf)

381

### Phew! Now let's grab Hillary's tweets.

In [131]:
tweets=[]
for page in tweepy.Cursor(api.user_timeline, screen_name='HillaryClinton').pages():
    tweets.append(page)

In [132]:
# Need to unpack tweets from lists in lists
tw = [status for tweet in tweets for status in tweet]

In [133]:
HC = []

for tweet in tw:
    HC.append(parse_tweets(tweet))
    

In [134]:
clintondf = pd.DataFrame(HC,columns = ['datetime', 'id', 'text', 'source'])

In [144]:
for i in range(1000):
    print clintondf.text[i]
    print 

Check out our new tool: Every time @realDonaldTrump tweets something offensive, automatically donate. https://t.co/6pWTY4BPhY

Trump reportedly asked this about nuclear weapons: "If we have them, why can't we use them?"

We can't let him beco… https://t.co/u3YYU5jbhP

Grateful to everyone who performed and turned out at #BroadwayForHillary. A great reminder about how important it i… https://t.co/BnGOv2trkB

Ryan talks about living with a disability—and befriending Hillary back in 1994: "She looks at me the way I always w… https://t.co/g0XeBjLbTT

The fate of the Supreme Court hinges on the choice we make in 21 days. Here's what's at stake: https://t.co/5fbqR2mDy2

Automatically donate every time Trump tweets something offensive with our new tool: https://t.co/XcTt51uKsd https://t.co/WPM1I9yeVv

RT @JoeBiden: Be one of the millions of Americans who vote before Election Day. Head to https://t.co/tu6W6eS9xR for more information. https…

"In the quiet moments backstage or at the end of a l

In [143]:
print clintondf.datetime.max()
print clintondf.datetime.min()

2016-10-19 02:07:32
2016-06-04 21:29:33


In [153]:
# Tweets by Hillary are indicated with a "-H" at the end.

sum(clintondf.text.apply(lambda s: s.endswith('-H')))

66

In [154]:
mask_H = clintondf.text.apply(lambda s: s.endswith('-H'))

In [158]:
clintondf = clintondf[mask_H]

In [181]:
clintondf = clintondf[clintondf.datetime>'2016-07-25 00:00:00']

In [182]:
clintondf = clintondf.reset_index(drop=True)

In [183]:
clintondf

datetime                  id  \
0  2016-10-13 18:45:25  786638992741638144   
1  2016-10-11 21:49:43  785960597934313472   
2  2016-10-11 21:42:22  785958746010681344   
3  2016-10-11 21:41:25  785958509502271488   
4  2016-10-09 18:03:44  785178953082363908   
5  2016-10-07 01:18:57  784201313257127936   
6  2016-10-04 23:14:20  783445177427955713   
7  2016-10-04 19:25:04  783387480418574336   
8  2016-10-03 21:05:23  783050338492379136   
9  2016-10-02 22:40:58  782712005719384064   
10 2016-09-29 20:00:57  781584571427090432   
11 2016-09-29 17:46:13  781550664631586816   
12 2016-09-27 21:35:28  780883582051123200   
13 2016-09-26 00:50:28  780207878229598208   
14 2016-09-24 23:49:22  779830115643617280   
15 2016-09-23 18:42:49  779390581369864192   
16 2016-09-21 14:41:06  778604976516001796   
17 2016-09-20 17:35:09  778286389448339457   
18 2016-09-15 15:26:40  776442117048700928   
19 2016-09-13 16:35:48  775734738464178176   
20 2016-09-13 13:10:31  775683074747199488   
21 2016-09-12 20:48:09  775435854466744320   
22 2016-09-12 20:46:58  775435555601543168   
23 2016-09-11 12:07:46  774942510582038529   
24 2016-09-09 15:42:30  774271773340729344   
25 2016-09-05 19:07:48  772873887214407680   
26 2016-09-01 01:21:39  771156031401521153   
27 2016-08-31 15:28:05  771006655316320260   
28 2016-08-28 16:18:29  769932174191648770   
29 2016-08-26 14:05:52  769174024396103680   
30 2016-08-25 20:38:47  768910514688815105   
31 2016-08-24 20:05:45  768539813913985024   
32 2016-08-24 18:02:14  768508732623970304   
33 2016-08-16 22:24:26  765675613034250240   
34 2016-08-15 01:26:05  764996549558534144   
35 2016-08-11 02:39:47  763565546512248832   
36 2016-08-08 15:50:59  762677497691566081   
37 2016-07-29 20:02:19  759116867730571265   
38 2016-07-29 01:50:02  758841983330725888   

                                                 text              source  
0   .@FLOTUS, I'm in awe. Thanks for putting into ...  Twitter Web Client  
1   G’mar chatima tova to those celebrating #YomKi...  Twitter Web Client  
2   If you’re able, chip in to help the families a...  Twitter Web Client  
3   I’m really concerned for North and South Carol...  Twitter Web Client  
4   I strongly condemn today's attack in Jerusalem...           TweetDeck  
5   @leedanielsent Thanks, Lee! Great to have you ...  Twitter for iPhone  
6   .@TimKaine's spent his whole life fighting for...  Twitter Web Client  
7   Concerned for the millions of families in Hurr...           TweetDeck  
8   It’s Filipino American History Month, so lets ...           TweetDeck  
9   L'Shana Tovah to everyone celebrating Rosh Has...           TweetDeck  
10  Glad to be back in Iowa today where it all beg...           TweetDeck  
11  The images from the train derailment in NJ are...           TweetDeck  
12  This election is too important to sit out. Go ...           TweetDeck  
13  Today, the sacrifices of Gold Star families we...           TweetDeck  
14  The new @NMAAHC is an overdue tribute to Afric...           TweetDeck  
15  Charlotte should release police video of the K...  Twitter Web Client  
16  Keith Lamont Scott. Terence Crutcher. Too many...  Twitter Web Client  
17  Another unarmed Black man was shot in a police...  Twitter Web Client  
18  Let's mark the start of Hispanic Heritage Mont...           TweetDeck  
19  As Muslims celebrate Eid al-Adha—a festival of...           TweetDeck  
20  The @NCAA is right to pull tournament games fr...  Twitter for iPhone  
21  Like anyone who’s ever been home sick from wor...  Twitter Web Client  
22  Thanks to everyone who’s reached out with well...  Twitter Web Client  
23  We'll never forget the horror of September 11,...           TweetDeck  
24  We can only confront a challenge like Zika whe...           TweetDeck  
25  Unions gave us higher wages, weekends, and the...           TweetDeck  
26  Trump just failed his first foreign test. Dipl...           TweetDeck  
27  .@GovChristie, vetoing a minimum wage inc

### Now we have clintondf & trumpdf
Trump has tweeted 381 times since his nomination; Clinton has tweeted 38 times since hers. (The power of 10...)

#### let's now parse the text, assign parts of speech, write tagged speech to df, then find out the most common adjectives used are

In [171]:
from textblob import TextBlob

In [179]:
text = TextBlob(trumpdf.text[0])

In [180]:
text.tags

[(u'@', u'JJ'),
 (u'THEREALMOGUL', u'NN'),
 (u'41', u'CD'),
 (u'%', u'NN'),
 (u'of', u'IN'),
 (u'American', u'JJ'),
 (u'voters', u'NNS'),
 (u'believe', u'VBP'),
 (u'the', u'DT'),
 (u'election', u'NN'),
 (u'could', u'MD'),
 (u'be', u'VB'),
 (u'stolen', u'VBN'),
 (u'from', u'IN'),
 (u'DonaldTrump', u'NNP'),
 (u'due', u'JJ'),
 (u'to', u'TO'),
 (u'widespread', u'JJ'),
 (u'voter', u'NN'),
 (u'fraud', u'NN'),
 (u'Politico', u'NN')]

[nltk_data] Downloading package brown to /Users/emily/nltk_data...
[nltk_data]   Package brown is already up-to-date!
[nltk_data] Downloading package punkt to /Users/emily/nltk_data...
[nltk_data]   Package punkt is already up-to-date!
[nltk_data] Downloading package wordnet to /Users/emily/nltk_data...
[nltk_data]   Unzipping corpora/wordnet.zip.
[nltk_data] Downloading package averaged_perceptron_tagger to
[nltk_data]     /Users/emily/nltk_data...
[nltk_data]   Unzipping taggers/averaged_perceptron_tagger.zip.
[nltk_data] Downloading package conll2000 to /Users/emily/nltk_data...
[nltk_data]   Package conll2000 is already up-to-date!
[nltk_data] Downloading package movie_reviews to
[nltk_data]     /Users/emily/nltk_data...
[nltk_data]   Unzipping corpora/movie_reviews.zip.
Finished.
